completed Comment System

elka1907 opened the pull request:
<img width="1300" height="347" alt="test" src="https://github.com/user-attachments/assets/74b71353-1cf4-4425-b373-5d5784e83a3a" />

elka1907 committed:
completed Comment System

acmwizard commented on src/test/java/faang/school/postservice/util/service/comment/CommentServiceImplTest.java:
```diff
@@ -0,0 +1,252 @@
+        return Comment.builder()
+                .id(id)
+                .content("Comment " + id)
+                .authorId(AUTHOR_ID + id)
```
AUTHOR_ID + id не нужно прибавлять id

elka1907 replied:
скорректировала

acmwizard commented on src/test/java/faang/school/postservice/util/service/comment/CommentServiceImplTest.java:
```diff
@@ -0,0 +1,252 @@
+            .authorId(AUTHOR_ID)
+            .post(testPost)
+            .createdAt(LocalDateTime.now().minusHours(1))
+            .updatedAt(LocalDateTime.now().minusHours(1))
```
updatedAt - не проверена логика 
createdAt - тоже

elka1907 replied:
добавила

acmwizard commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        Comment comment = commentMapper.toEntity(createCommentDto);
+        comment.setPost(post);
+
+        Comment savedComment = commentRepository.save(comment);
```
createdAt-?
updatedAt-?

elka1907 replied:
настроено в model comment-    есть @CreationTimestamp которая позволит спрингу поставить дату создания

acmwizard commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+            throw new IllegalArgumentException("Comment does not belong to post with id: " + postId);
+        }
+
+        commentMapper.updateEntity(updateCommentDto, newComment);
```
createdAt-?
updatedAt-?

elka1907 replied:
настроено в model comment-  есть @UpdateTimestamp которая позволит спрингу обновить дату

acmwizard commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+            throw new IllegalArgumentException("Comment does not belong to post with id: " + postId);
+        }
+
+        commentRepository.deleteById(commentId);
```
updatedAt-?

elka1907 replied:
запись удаляется и дата тоже

acmwizard commented on src/main/java/faang/school/postservice/mapper/comment/CommentMapper.java:
```diff
@@ -0,0 +1,36 @@
+        unmappedTargetPolicy = org.mapstruct.ReportingPolicy.IGNORE)
+public interface CommentMapper {
+
+    @Mapping(source = "post.id", target = "postId")
```
"post.id" или просто id

elka1907 replied:
В маппинге нужно использовать source = "post.id", а не просто "id", потому что "id" - это id самого комментария

acmwizard commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class CommentServiceImpl implements CommentService {
```
public Post getPostEntityById(long id) {  - где этот метод тут нужен?

elka1907 replied:
я просила, чтобы ты у себя добавил данный метод, т.к. я его буду использовать в createComment, getAllComments для проверки поста

elka1907 commented:
не вижу какие ошибки в коде.

elka1907 committed:
скорректировала добавила equals, в тесты добавила проверку на createdAt  и updatedAt

acmwizard commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,68 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/posts/{postId}/comments")
+@Validated
```
Есть вопросы к Validated, вроде как не должны использовать

Соглашение по валидации:
Используем аннотации валидации (@NotBlank, ....)  только к ДТО для запросов.
К ДТО ответов не стоит добавлять.
Для включения аннотаций используем @Valid (@Validated не нужно)

https://discord.com/channels/1102662981743030334/1126205623265140736/1428389069666123906

elka1907 replied:
убрала

acmwizard commented:
:thumbsup:

elka1907 committed:
validated убрала

elka1907 commented:
Костя, моя задачка завязана с системой постов, когда у Андрея задача будет merged, мне надо в CommentServiceImpl скорректировать некоторые пункты на PostService: 
1. private final PostRepository postRepository; // заменить на PostService postService;
2. в getAllComments добавить postService.getPostEntityById(postId);
3. в createComment: Post post = postRepository.findById(postId) //заменить на = postService.getPostEntityById(postId);
Спасибо)

SnurnicynKonstantin commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,66 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/posts/{postId}/comments")
```
Лучше здесь оставить /comments. А уже в ендпоинтах /post/{postId} использовать для операций

elka1907 replied:
скорректировала

SnurnicynKonstantin commented on src/main/java/faang/school/postservice/dto/comment/CreateCommentDto.java:
```diff
@@ -0,0 +1,14 @@
+
+public record CreateCommentDto(
+        @NotNull(message = "Author ID cannot be null")
+        Long authorId,
```
Автора из UserContext надо брать

elka1907 replied:
скорректировала

SnurnicynKonstantin commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+            throw new IllegalArgumentException("Comment content cannot be blank");
+        }
+
+        if (content.length() > 4096) {
```
4096 в константу поместить бы

elka1907 replied:
добавила

SnurnicynKonstantin commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,66 @@
+    @DeleteMapping("/{commentId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteComment(
+            @PathVariable long postId,
```
postId зачем нам здесь?

SnurnicynKonstantin commented on src/main/java/faang/school/postservice/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+                .orElseThrow(() -> new IllegalArgumentException("Post not found with id: " + postId));
+
+        try {
+            userServiceClient.getUser(createCommentDto.authorId());
```
А ты проверяла, ответы приходят от юзер сервиса? 
Сейчас вижу, что они уходят на `локалхост:8080/users/{userId}`, но у нас-то там `локалхост:8080/api/v1/users/{userId}`. По этому надо добавить /api/v1 клиент.
<img width="566" height="198" alt="image" src="https://github.com/user-attachments/assets/9926945c-31d8-4e76-9b01-ce3071ed7bda" />

Полюс FeignException это когда достучаться до сервиса невозможно выпадет. Тебе надо слушать ответ метода getUser(). Если там код 404, то ошибка. Если 200, то идем дальше.

elka1907 replied:
добавила в FeignException и скорректировала сервис

elka1907 committed:
скорректировала dto, на USER_ID, путь в URL

# Conflicts:
#	src/main/java/faang/school/postservice/client/UserServiceClient.java
